zen_document_parser/base.py

```diff
@@ -0,0 +1,231 @@
+from enum import IntEnum
+from pdfquery import PDFQuery
+from pdfminer.pdfdocument import PDFSyntaxError
+
+from zen_document_parser.exceptions import InvalidPDFError, UnknownVariantError
+
+
+class DocFieldType(IntEnum):
+    TEXT = 1
+    NUMBER = 2
+    CUSTOM = 3  # TODO: Forget this and have 'type' take a callable instead?
+
+
+class DocField(object):
+    def __init__(self, bbox, type=DocFieldType.TEXT, required=False,
+                 description=None):
+        self.bbox = bbox
+        self.type = type
+        self.required = required
+        self.description = description
+
+
+class DocVariantMeta(type):
+    """Metaclass for DocVariant. Allows defining a 'schema' based on DocFields
+    on a DocVariant subclass (similar to Django models).
+    """
+
+    def __new__(cls, cls_name, bases, members):
+
+        # To ensure proper handling of fields defined in parents
+        parents = [b for b in bases if isinstance(b, DocVariantMeta)]
+
+        module = members.pop('__module__')
+        new_class = super(DocVariantMeta, cls).__new__(
+            cls, cls_name, bases, {'__module__': module}
+        )
+
+        _meta = {'field_info': {}}
+        setattr(new_class, '_meta', _meta)
+        meta_field_info = _meta['field_info']
+
+        misc_members = []
+        doc_field_members = []
+
+        # Collect doc fields
+        for field_name, field_obj in members.iteritems():
+            if isinstance(field_obj, DocField):
+                doc_field_members.append((field_name, field_obj))
+            else:
+                misc_members.append((field_name, field_obj))
+
+        # Also get doc fields from parents
+        for parent in parents:
+            doc_field_members.extend(parent._meta['field_info'].iteritems())
+
+        # Put misc stuff into class
+        for member_name, member_value in misc_members:
+            setattr(new_class, member_name, member_value)
+
+        # Process doc fields
+        for field_name, field_obj in doc_field_members:
+            # Populate field info in meta
+            meta_field_info[field_name] = field_obj
+
+            # TODO: Fix initial value. This is a placeholder.
+            setattr(new_class, field_name, field_name)
+
+        return new_class
+
+
+class DocVariant(object):
+
+    __metaclass__ = DocVariantMeta
+
+    # The fields that are loaded up for use inside check_for_match(). This is
+    # an optimization that avoids us needing to extract every field of the
+    # document just to perform a test using a few fields.
+    # TODO: Find a way that doesn't require specifying this manually.
+    test_fields = []
+
+    def __init__(self, file):
+        """
+        Args:
+            file (PDFQuery object): A **loaded** PDFQuery object that will
+                be used internally to extract fields.
+        """
+        self._file = file
```

Comment:
Ah, you mean it's a built in function. I didn't know that.
https://docs.python.org/2/library/functions.html#file

Though it seems to be deprecated in Python 2.x and removed in 3.x.